File: plugins/ainb-fleet/bridge/ainb_phone_bridge/secrets.py
# ABOUTME: Secret resolver for the phone bridge config.
#
# Mirrors agent-deck's embedded-template `_resolve_secret` contract (the verified
# one, NOT the older standalone bridge.py which only accepts plain strings):
#   "$ENV_VAR"     -> os.environ.get("ENV_VAR", "")  (warns if unset)
#   "${ENV_VAR}"   -> same branch (strips both '$' and surrounding '{}')
#   "keychain:svc" -> `security find-generic-password -s svc -w` (macOS only)
#   "plain-string" -> returned unchanged
#
# A secret is NEVER passed on argv into the daemon — the bridge reads the raw
# value from config and resolves it here, in-process, at startup.

from __future__ import annotations

import logging
import subprocess

log = logging.getLogger("ainb.bridge.secrets")

KEYCHAIN_PREFIX = "keychain:"
# `security` can hang on a locked keychain prompt; bound it.
_KEYCHAIN_TIMEOUT_S = 5


def resolve_secret(value: str) -> str:
    """Resolve a config secret reference to its plaintext value.

    Pure aside from reading ``os.environ`` and (for keychain refs) shelling out
    to ``/usr/bin/security``. Unknown / missing references resolve to ``""`` with
    a warning rather than raising — the bridge then fails loudly on an empty
    token at startup, which is a clearer error than a stack trace deep in here.
    """
    import os

    if not isinstance(value, str):
        return ""

    raw = value.strip()
    if not raw:
        return ""

    if raw.startswith("$"):
        # Handles both "$VAR" and "${VAR}" — strip the leading '$' and any
        # surrounding braces, then read the environment (default "").
        name = raw.lstrip("$").strip("{}")
        resolved = os.environ.get(name, "")
        if not resolved:
            log.warning("env var %s referenced by config is unset/empty", name)
        return resolved

    if raw.startswith(KEYCHAIN_PREFIX):
        service = raw[len(KEYCHAIN_PREFIX) :]
        if not service:
            log.warning("keychain reference has no service name")
            return ""
        try:
            out = subprocess.run(
                ["/usr/bin/security", "find-generic-password", "-s", service, "-w"],
                capture_output=True,
                text=True,
                timeout=_KEYCHAIN_TIMEOUT_S,
                check=False,
            )
        except (OSError, subprocess.TimeoutExpired) as exc:
            log.warning("keychain lookup for %s failed: %s", service, exc)
            return ""
        if out.returncode != 0:
            log.warning("keychain has no item for service %s", service)
            return ""
        return out.stdout.strip()

    # Plain literal — returned as-is.
    return raw

File: ainb-tui/crates/ainb-core/src/fleet/bridge/secrets.rs
// ABOUTME: Secret resolver for the native phone-bridge config.
//
// Ported from the Python `ainb_phone_bridge.secrets` (the verified contract):
//   "$ENV_VAR"     -> std::env::var("ENV_VAR").unwrap_or_default()  (warns if unset)
//   "${ENV_VAR}"   -> same branch (strips '$' and surrounding '{}')
//   "keychain:svc" -> `security find-generic-password -s svc -w` (macOS only)
//   "plain-string" -> returned unchanged
//
// A secret is NEVER passed on argv into the daemon — the bridge reads the raw
// reference from config and resolves it here, in-process, at startup. Unknown /
// missing references resolve to "" with a warning rather than erroring, so the
// bridge fails loudly on an empty token at startup (a clearer error than a deep
// stack trace).

use std::process::Command;
use std::time::Duration;

const KEYCHAIN_PREFIX: &str = "keychain:";
/// `security` can hang on a locked-keychain prompt; bound it. (Enforced via a
/// watchdog thread because `std::process` has no built-in timeout.)
const KEYCHAIN_TIMEOUT: Duration = Duration::from_secs(5);

/// Resolve a config secret reference to its plaintext value.
///
/// Pure aside from reading the process environment and (for keychain refs)
/// shelling out to `/usr/bin/security`.
#[must_use]
pub fn resolve_secret(value: &str) -> String {
    resolve_secret_with(value, |name| std::env::var(name).ok())
}

/// Inner resolver with an injectable env lookup, so the `$VAR` / `${VAR}` branch
/// is unit-testable without mutating the process environment (the crate forbids
/// `unsafe`, and `std::env::set_var` is process-global / racy anyway).
fn resolve_secret_with(value: &str, env: impl Fn(&str) -> Option<String>) -> String {
    let raw = value.trim();
    if raw.is_empty() {
        return String::new();
    }

    if let Some(rest) = raw.strip_prefix('$') {
        // Handles both "$VAR" and "${VAR}" — strip the leading '$' and any
        // surrounding braces, then read the environment (default "").
        let name = rest.trim_matches(|c| c == '{' || c == '}');
        let resolved = env(name).unwrap_or_default();
        if resolved.is_empty() {
            tracing::warn!(
                env_var = name,
                "env var referenced by bridge config is unset/empty"
            );
        }
        return resolved;
    }

    if let Some(service) = raw.strip_prefix(KEYCHAIN_PREFIX) {
        if service.is_empty() {
            tracing::warn!("keychain reference has no service name");
            return String::new();
        }
        return resolve_keychain(service);
    }

    // Plain literal — returned as-is.
    raw.to_string()
}

/// macOS keychain lookup, bounded by a watchdog so a locked-keychain prompt
/// can't hang the bridge at startup.
fn resolve_keychain(service: &str) -> String {
    let service_owned = service.to_string();
    let (tx, rx) = std::sync::mpsc::channel();
    std::thread::spawn(move || {
        let out = Command::new("/usr/bin/security")
            .args(["find-generic-password", "-s", &service_owned, "-w"])
            .output();
        let _ = tx.send(out);
    });

    match rx.recv_timeout(KEYCHAIN_TIMEOUT) {
        Ok(Ok(out)) if out.status.success() => {
            String::from_utf8_lossy(&out.stdout).trim().to_string()
        }
        Ok(Ok(_)) => {
            tracing::warn!(service, "keychain has no item for service");
            String::new()
        }
        Ok(Err(e)) => {
            tracing::warn!(service, error = %e, "keychain lookup failed");
            String::new()
        }
        Err(_) => {
            tracing::warn!(service, "keychain lookup timed out");
            String::new()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn plain_literal_passes_through() {
        assert_eq!(resolve_secret("xoxb-123"), "xoxb-123");
    }

    #[test]
    fn whitespace_only_is_empty() {
        assert_eq!(resolve_secret("   "), "");
        assert_eq!(resolve_secret(""), "");
    }

    #[test]
    fn dollar_env_is_read() {
        // Inject a fake env so the test never mutates the process environment.
        let env = |name: &str| (name == "MYVAR").then(|| "resolved-token".to_string());
        assert_eq!(resolve_secret_with("$MYVAR", &env), "resolved-token");
        assert_eq!(resolve_secret_with("${MYVAR}", &env), "resolved-token");
    }

    #[test]
    fn missing_env_resolves_to_empty() {
        let env = |_: &str| None;
        assert_eq!(
            resolve_secret_with("$AINB_BRIDGE_DEFINITELY_UNSET_XYZ", &env),
            ""
        );
    }

    #[test]
    fn empty_keychain_service_is_empty() {
        assert_eq!(resolve_secret("keychain:"), "");
    }
}
